src/proxy/utils.ts

```diff
@@ -1,125 +1,53 @@
+import { IProxyRouteConfig, IStreamOpaque } from './types'
 
-export const buildHostUrl = (): string | null => {
-	const host = config.get('proxy.hosts.relayProxy') as string | null
-
-	if (host) {
-		if (host.startsWith('https://')) {
-			return host
-		}
-		return `https://${host}`
-	}
-
-	return null
-}
-
-export const bufferToJson = (buffer: Buffer): unknown => {
-	try {
-		return JSON.parse(buffer.toString('utf-8'))
-	} catch (err) {
-		logger.error(
-			{ err, buffer: buffer.toString('utf-8') },
-			'Failed to parse json'
-		)
-		return null
+export const buildUrl = (route: IProxyRouteConfig, reqUrl: string): string => {
+	if (route.stripPrefix) {
+		return `${route.targetHost}${reqUrl.substring(route.prefix.length)}`
 	}
-}
-
-const jsonToBuffer = (json: unknown): Buffer => {
-	try {
-		return Buffer.from(JSON.stringify(json))
-	} catch (err) {
-		logger.error({ err, json }, 'Failed to stringify json')
-		return Buffer.from('{ "error": "Failed to process upstream response" }')
-	}
-}
-
-export const compressJson = async (json: unknown): Promise<Buffer> => {
-	return new Promise((resolve, reject) => {
-		zlib.gzip(jsonToBuffer(json), (err, buffer) => {
-			if (err) {
-				logger.error({ err, json }, 'Failed to compress json')
-				reject(err)
-			}
-			resolve(buffer)
-		})
-	})
-}
-
-export const decompressResponse = async (stream: Readable): Promise<Buffer> => {
-	const gunzip = zlib.createGunzip()
-	stream.pipe(gunzip)
-	return new Promise((resolve) => {
-		const buffers: Buffer[] = []
-
-		gunzip
-			.on('data', (chunk: Buffer) => {
-				buffers.push(chunk)
-			})
-			.on('end', () => {
-				resolve(Buffer.concat(buffers))
-			})
-			.on('error', (err: Error) => {
-				logger.error({ err }, 'Failed to decompress response')
-				resolve(
-					Buffer.from('{ "error": "Failed to decompress response" }')
-				)
-			})
-	})
-}
-
-const stripPrefix = (reqUrl: string, prefix?: string): string => {
-	if (prefix && reqUrl.startsWith(prefix)) {
-		return reqUrl.substring(prefix.length)
-	}
-
-	return reqUrl
+	return `${route.targetHost}${reqUrl}`
 }
 
 export const prepareHeaders = (
+	host: string,
 	originalHeaders: IncomingHttpHeaders,
 	remoteAddress?: string
-): IncomingHttpHeaders => {
-	return {
-		...originalHeaders,
-		'X-Forwarded-For': originalHeaders['x-forwarded-for'] ?? remoteAddress,
-		// Next line bypasses: [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames.
-		host: '',
-	}
-}
-
-export const buildUrl = (route: IProxyRouteConfig, url: string): string => {
-	return `${route.targetHost}${stripPrefix(url, route.proxyPrefix)}`
-}
+): IncomingHttpHeaders => ({
+	...originalHeaders,
+	'x-forwarded-for': originalHeaders['x-forwarded-for'] ?? remoteAddress,
+	'x-forwarded-host': host,
+	host,
+})
 
 export const pipeResponse = async (
+	routeTargetHost: string,
 	url: string,
 	req: IncomingMessage,
 	res: ServerResponse
 ): Promise<void> => {
+	const host = routeTargetHost.split('https://')[1] ?? ''
```

Comment:
I would prefer `routeTargetHost.replace('https://', '')`. It just create one string instead of an array with two strings.

Reply:
Good point👍🏽